client/authority-discovery/src/worker.rs

```diff
@@ -495,51 +482,70 @@ where
 		let remote_addresses: Vec<Multiaddr> = values
 			.into_iter()
 			.map(|(_k, v)| {
-				let schema::SignedAuthorityAddresses { signature, addresses } =
-					schema::SignedAuthorityAddresses::decode(v.as_slice())
+				let schema::SignedAuthorityRecord { record, auth_signature, peer_signature } =
+					schema::SignedAuthorityRecord::decode(v.as_slice())
 						.map_err(Error::DecodingProto)?;
 
-				let signature = AuthoritySignature::decode(&mut &signature[..])
+				let auth_signature = AuthoritySignature::decode(&mut &auth_signature[..])
 					.map_err(Error::EncodingDecodingScale)?;
 
-				if !AuthorityPair::verify(&signature, &addresses, &authority_id) {
+				if !AuthorityPair::verify(&auth_signature, &record, &authority_id) {
 					return Err(Error::VerifyingDhtPayload)
 				}
 
-				let addresses = schema::AuthorityAddresses::decode(addresses.as_slice())
-					.map(|a| a.addresses)
-					.map_err(Error::DecodingProto)?
-					.into_iter()
-					.map(|a| a.try_into())
-					.collect::<std::result::Result<_, _>>()
-					.map_err(Error::ParsingMultiaddress)?;
-
+				let mut addresses: Vec<Multiaddr> =
+					schema::AuthorityRecord::decode(record.as_slice())
+						.map(|a| a.addresses)
+						.map_err(Error::DecodingProto)?
+						.into_iter()
+						.map(|a| a.try_into())
+						.collect::<std::result::Result<_, _>>()
+						.map_err(Error::ParsingMultiaddress)?;
+
+				let get_peer_id = |a: &Multiaddr| match a.iter().last() {
+					Some(multiaddr::Protocol::P2p(key)) =>
+						PeerId::from_multihash(key).map_err(|_| ()),
+					_ => return Err(()),
+				};
+
+				// Ignore [`Multiaddr`]s without [`PeerId`] and own addresses.
+				let addresses: Vec<Multiaddr> = addresses
+					.drain(..)
+					.filter(|a| match get_peer_id(a) {
+						Ok(peer_id) => peer_id != local_peer_id,
+						Err(_) => false,
+					})
+					.collect();
+
+				let remote_peer_id = single(addresses.iter().map(get_peer_id))
+					.map_err(|_| Error::ReceivingDhtValueFoundEventWithDifferentPeerIds)? // different peer_id in records
+					.ok_or(Error::ReceivingDhtValueFoundEventWithNoPeerIds)? // no records
+					.map_err(|()| Error::ReceivingDhtValueFoundEventWithNoPeerIds)?; // no peer_id in any records
+
+				if let Some(peer_signature) = peer_signature {
+					let public_key = libp2p::identity::PublicKey::from_protobuf_encoding(
+						&peer_signature.public_key,
+					)
+					.map_err(|e| Error::ParsingLibp2pIdentity(e))?;
+					let peer_id_in_signature = public_key.clone().to_peer_id();
+					if remote_peer_id != peer_id_in_signature ||
+						!public_key.verify(&record, &peer_signature.signature)
+					{
+						return Err(Error::VerifyingDhtPayload)
+					}
+				} else if self.strict_record_validation {
+					return Err(Error::MissingPeerIdSignature)
+				} else {
+					warn!(
+						target: LOG_TARGET,
+						"Received unsigned authority discovery record from {}", authority_id
+					);
```

Comment:
This should not be a warning.

Reply:
Made it an `info!(...)` then

Reply:
No also not a info. What should a node operator do with this log line? It is only spamming it logs, please use debug.